### DISCUSSION TITLE
Django rest API project - Need help to emit event from a class based view

### DISCUSSION THREAD

**Discussion by tutoojoe** [miguelgrinberg/python-socketio#926] [Q&A]
I am working on a Django REST framework project with ReactJS frontend.
I wanted to use a real time update functionality in my project with Python-SocketIO.
The connection between Server and client is successfully established.
_The emit functionality is working fine with functional components._

In my  project, the views I have written are class based views, using model serializers, and hence I couldn't wrap the view using the decorator.
Below is one of the model, serializer and its view.

### **model**
```
class ProductGroup(models.Model):
    product_group = models.CharField(_("Product group"), max_length=20)

    def __str__(self):
        return self.product_group
```

### **serializer (using model serializer)**
```
class ProductGroupSerializer(serializers.ModelSerializer):
    class Meta:
        model = ProductGroup
        fields = "__all__"
```
### **view - class based, created with generic class from rest framework**
```

class ProductGroupListCreateAPIView(generics.ListCreateAPIView):
    serializer_class = ProductGroupSerializer

    def get(self, request, format=None):
         product_groups = ProductGroup.objects.all()
        serializer = self.serializer_class(product_groups, many=True)
        return Response(serializer.data)
```
I got stuck here as I don't have enough knowledge to implement the emit event inside this view.
Please guide me on wrapping the view with event and emiting events.
I am new to programming and request your help on this.
Thanks in advance.

Thanks in advance.

**Comment by miguelgrinberg** [comment]
The Socket.IO event handlers are completely independent of your Django views. You can create them as decorated functions, or if you prefer a class-based style you can use a class-based namespace. There are examples of both approaches in the documentation.

To emit an event to the client, you use the `sio.emit()` function. You can use this from a Django view without any issues.

**Reply by tutoojoe** [reply]
@miguelgrinberg thanks for the reply.

I am not sure whether the class based Namespace will work in my view, I will check it again.

I will elaborate a bit more on my scenario. The view I wrote above is a generic view, extended from another class.
The view accepts 2 http methods, POST - for creating an object and GET for retrieving the complete list of objects.  But this is being managed by DRF based on the serializer and queryset. 
My idea was to send an emit signal every time an object is created and do something in the front end based on this signal.
(Similarly PUT, PATCH, DELETE method in other views also).
I assume that, I may have to change the view from 'generic view' to normal API view so that I will get a better grip on the http methods. 

If you have any better suggestion, could you please advise me? (if possible with example)

Thanks in advance.

**Reply by miguelgrinberg** [reply]
I think you are over-complicating this. If all you want to do in your back end is to send signals to the front end, then you do not need to handle any events (except maybe the `connect` event).

So there's no point in using a class-based namespace, you can just use a decorated event handler for `connect` (if you need this), and then to emit you can just call the emit function from any views that need to send signals to the front end.

There is a Django example in this repository if you want to see how the integration can be achieved.

**Reply by tutoojoe** [reply]
I have tried it in the below way and it worked perfectly.
I have changed the `view` part of my program as below in normal `APIView`
```
class ProductGroupListCreateAPIView(APIView):
    """
    List all snippets, or create a new snippet.
    """
    def get(self, request, format=None):
        product_groups = ProductGroup.objects.all()
        serializer = ProductGroupSerializer(product_groups, many=True)
        return Response(serializer.data)

    @sio.event
    def post(self, request, format=None):
        serializer = ProductGroupSerializer(data=request.data)
        if serializer.is_valid():
            serializer.save()
            sio.emit('product_group_added', {'msg': 'product group added'})
            return Response(serializer.data, status=status.HTTP_201_CREATED)
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

This worked, and I am changing the other `generic views` to normal `APIViews` so that I can implement the functionality.

Thanks @miguelgrinberg  for your time and for the replies to my submission.

**Reply by miguelgrinberg** [reply]
Remove the `@sio.event` from your view. That is not necessary.

**Reply by tutoojoe** [reply]
Removed `@sio.event`, and is working fine.
Thanks for advising correction.
